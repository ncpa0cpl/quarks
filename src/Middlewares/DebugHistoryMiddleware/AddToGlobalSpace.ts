export function addToGlobalSpace(o: object) {
  if (window) {
    Object.assign(window, o);
  } else if (global && global.window) {
    Object.assign(global.window, o);
  } else if (global) {
    Object.assign(global, o);
  }
}
