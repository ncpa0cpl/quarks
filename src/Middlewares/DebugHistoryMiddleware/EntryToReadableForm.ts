import { DateTime } from "luxon";
import type { RecordValue } from "../../Types";
import { hasKey } from "../../Utilities/GeneralPurposeUtilities";
import type {
  HistoricalState,
  HistoryPropertiesKeys,
  QuarkStateChangeHistoricalEntry,
} from "./Types/TrackedQuark";

const PROPERTIES_FRIENDLY_NAMES_MAP: Record<HistoryPropertiesKeys, string> = {
  value: "Value:",
  type: "Value Type:",
  change: "Value Change is:",
  initialState: "Previous state value:",
  dispatchedUpdate: "Dispatched state value:",
  name: "Quark Name:",
  source: "Update Source:",
  stackTrace: "Stack Trace:",
  stateAfterUpdate: "Next state value:",
  time: "Timestamp:",
  isCanceled: "Canceled:",
};

const columns: HistoryPropertiesKeys[] = [
  "initialState",
  "dispatchedUpdate",
  "stateAfterUpdate",
  "isCanceled",
  "change",
  "source",
  "time",
  "stackTrace",
];

function stringifyIfObject(v: unknown) {
  if (typeof v === "object") {
    if (v instanceof Promise) return "Promise {}";
    if (v !== null) return JSON.stringify(v);
    return "null";
  }
  return v;
}

function parseObjectValue(obj: HistoricalState) {
  if ("type" in obj && "value" in obj)
    return `Type: ${obj.type}; Value: [${stringifyIfObject(obj.value)}]`;

  return `Type: Value; Value: [${stringifyIfObject(obj)}]`;
}

export function entryToReadableForm(entry: QuarkStateChangeHistoricalEntry) {
  const columnValues = columns.map(
    (propertyName: HistoryPropertiesKeys): [string, unknown] => {
      const friendlyName = PROPERTIES_FRIENDLY_NAMES_MAP[propertyName];
      const value: RecordValue<QuarkStateChangeHistoricalEntry> = hasKey(
        entry,
        propertyName,
      )
        ? entry[propertyName]
        : undefined;

      if (typeof value === "object") {
        return [friendlyName, parseObjectValue(value)];
      }

      if (propertyName === "time") {
        return [friendlyName, DateTime.fromMillis(value as number).toISO()];
      }

      if (propertyName === "isCanceled") {
        return [friendlyName, !!value];
      }

      return [friendlyName, value];
    },
  );

  return Object.fromEntries(columnValues);
}
