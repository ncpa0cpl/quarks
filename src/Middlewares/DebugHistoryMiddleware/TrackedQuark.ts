import { DateTime } from "luxon";
import type {
  AddHistoryEntryParam,
  QuarkStateChangeHistoricalEntry,
  TrackedQuark,
} from "./Types/TrackedQuark";

export function createTrackedQuark(name: string): TrackedQuark {
  const stateChangeHistory: QuarkStateChangeHistoricalEntry[] = [];

  const addHistoryEntry = (entry: AddHistoryEntryParam) => {
    const change = ["Promise", "Generator"].includes(entry.dispatchedUpdate.type)
      ? "Postponed"
      : "Immediate";

    stateChangeHistory.push({
      ...entry,
      time: DateTime.now().toMillis(),
      change,
      stateAfterUpdate:
        change === "Postponed"
          ? entry.initialState.value
          : typeof entry.dispatchedUpdate.value === "function"
          ? entry.dispatchedUpdate.value(entry.initialState.value)
          : entry.dispatchedUpdate.value,
    });
  };

  const clear = () => {
    stateChangeHistory.splice(0);
  };

  return {
    name,
    stateChangeHistory,
    addHistoryEntry,
    clear,
  };
}
