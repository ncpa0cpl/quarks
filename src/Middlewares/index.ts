export { createCatchMiddleware } from "./CatchMiddleware";
export { createDebugHistoryMiddleware } from "./DebugHistoryMiddleware";
