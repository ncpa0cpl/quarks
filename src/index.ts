export * from "./Middlewares";
export { quark } from "./Quark";
export { composeSelectors } from "./SelectorCompose";
export * from "./Types/index";
export { hydrateQuarks, serializeQuarks } from "./Utilities";
export { CancelUpdate } from "./Utilities/CancelUpdate";
export {
  addGlobalQuarkMiddleware,
  getGlobalQuarkMiddlewares,
  setGlobalQuarkMiddlewares,
} from "./Utilities/GlobalMiddlewares";
