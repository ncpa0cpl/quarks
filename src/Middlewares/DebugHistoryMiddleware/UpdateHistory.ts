import { createTrackedQuark } from "./TrackedQuark";
import type { TrackedQuark } from "./Types/TrackedQuark";

function createStateUpdateHistory() {
  const trackedQuarks: TrackedQuark[] = [];

  const track = (name: string) => {
    const quarkTrack = createTrackedQuark(name);
    trackedQuarks.push(quarkTrack);
    return quarkTrack;
  };

  const showHistory = () => {
    return Object.fromEntries(
      trackedQuarks.map((trackedQuark, index) => {
        return [`${index}_${trackedQuark.name}`, trackedQuark.stateChangeHistory];
      })
    );
  };

  return { track, showHistory };
}

export const StateUpdateHistory = createStateUpdateHistory();

if (global.window) {
  Object.assign(global.window, { __quark_history_tracker__: StateUpdateHistory });
} else if (global) {
  Object.assign(global, { __quark_history_tracker__: StateUpdateHistory });
}
