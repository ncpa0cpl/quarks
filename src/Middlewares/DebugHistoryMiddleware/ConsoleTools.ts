import { DateTime } from "luxon";
import { hasKey } from "../../Utilities/GeneralPurposeUtilities";
import { addToGlobalSpace } from "./AddToGlobalSpace";
import type { HistoricalState, HistoryPropertiesKeys } from "./Types/TrackedQuark";
import { StateUpdateHistory } from "./UpdateHistory";

const PROPERTIES_FRIENDLY_NAMES_MAP: Record<HistoryPropertiesKeys, string> = {
  value: "Value:",
  type: "Value Type:",
  change: "Value Change is:",
  initialState: "Quark State Value before update:",
  dispatchedUpdate: "Dispatched Value:",
  name: "Quark Name:",
  source: "Update Source:",
  stackTrace: "Stack Trace:",
  stateAfterUpdate: "State Value after applying this update:",
  time: "Timestamp:",
};

function parseHistoricalStateToString(obj: HistoricalState) {
  return `Type: ${obj.type}; Value: [${obj.value}]`;
}

function printQuarkHistory(options?: { name?: string; showLast?: number }) {
  const { showLast = 16, name = undefined } = options ?? {};

  const history = StateUpdateHistory;

  const quarksHistories = history.getHistory();

  const columns: HistoryPropertiesKeys[] = [
    "initialState",
    "dispatchedUpdate",
    "stateAfterUpdate",
    "change",
    "source",
    "time",
    "stackTrace",
  ];

  for (const quarkHistory of quarksHistories) {
    if (!name || quarkHistory.name === name) {
      console.group(quarkHistory.name);

      const table = quarkHistory.stateChangeHistory
        .reverse()
        .slice(0, showLast)
        .map((entry) => {
          const columnValues = columns.map((propertyName): [string, unknown] => {
            const friendlyName = PROPERTIES_FRIENDLY_NAMES_MAP[propertyName];
            const value = hasKey(entry, propertyName)
              ? entry[propertyName]
              : undefined;

            if (typeof value === "object") {
              return [
                friendlyName,
                parseHistoricalStateToString(value as HistoricalState),
              ];
            }

            if (propertyName === "time") {
              return [friendlyName, DateTime.fromMillis(value as number).toISO()];
            }

            return [friendlyName, value];
          });

          return Object.fromEntries(columnValues);
        })
        .reverse();

      console.table(table);
      console.groupEnd();
    }
  }
}

export function initConsoleTools() {
  addToGlobalSpace({ printQuarkHistory });
}
