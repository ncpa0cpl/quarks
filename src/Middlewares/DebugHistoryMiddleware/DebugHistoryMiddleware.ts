import { cloneDeep as _cloneDeep } from "lodash";
import type { QuarkMiddleware } from "../../Types";
import { extractIsPromiseCanceled } from "../../Utilities/StateUpdates/AsyncUpdates";
import { getStateUpdateHistory } from "./UpdateHistory";

function getValueType(val: any) {
  if (val instanceof Promise) return "Promise";
  if (typeof val === "function") return "Generator";
  return "Value";
}

function cloneDeep<T>(v: T): T {
  if (v instanceof Promise || typeof v === "function") return v;
  return _cloneDeep(v);
}

export function createDebugHistoryMiddleware(options: {
  name: string;
  trace?: boolean;
  realTimeLogging?: boolean;
  useTablePrint?: boolean;
}): QuarkMiddleware<any, undefined> {
  const {
    name,
    trace = true,
    realTimeLogging = false,
    useTablePrint = true,
  } = options;
  const StateUpdateHistory = getStateUpdateHistory();
  const quarkHistoryTracker = StateUpdateHistory.track({
    name,
    realTimeLogging,
    useTablePrint,
  });
  return (getState, newValue, resume, _, type) => {
    switch (type) {
      case "sync": {
        const stackTrace = trace
          ? new Error().stack?.replace(/$Error\n\sat/gi, "Called from")
          : undefined;
        quarkHistoryTracker.addHistoryEntry({
          source: "Sync-Dispatch",
          stackTrace,
          initialState: {
            type: "Value",
            value: cloneDeep(getState()),
          },
          dispatchedUpdate: {
            type: getValueType(newValue),
            value: cloneDeep(newValue),
          },
        });
        break;
      }
      case "async": {
        quarkHistoryTracker.addHistoryEntry({
          source: "Async-Dispatch",
          stackTrace: undefined,
          initialState: {
            type: "Value",
            value: cloneDeep(getState()),
          },
          dispatchedUpdate: {
            type: getValueType(newValue),
            value: cloneDeep(newValue),
          },
        });
        break;
      }
    }

    if (newValue instanceof Promise) {
      newValue
        .then((v) => {
          const hasBeenCanceled = extractIsPromiseCanceled(newValue);
          if (hasBeenCanceled) {
            quarkHistoryTracker.addHistoryEntry({
              dispatchedUpdate: {
                type: getValueType(v),
                value: cloneDeep(v),
              },
              initialState: {
                type: "Value",
                value: getState(),
              },
              source: "Async-Dispatch",
              stackTrace: undefined,
              isCanceled: true,
            });
          }
        })
        .catch((e) => {});
    }

    return resume(newValue);
  };
}
