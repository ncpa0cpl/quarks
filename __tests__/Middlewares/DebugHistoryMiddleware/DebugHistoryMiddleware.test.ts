import { DateTime } from "luxon";
import { createDebugHistoryMiddleware, quark } from "../../../src";
import { getStateUpdateHistory } from "../../../src/Middlewares/DebugHistoryMiddleware/UpdateHistory";
import { sleep } from "../../helpers";

declare const global: {
  __quark_history_tracker__: ReturnType<typeof getStateUpdateHistory>;
};

function* timeGenerator(): Generator<number, number, unknown> {
  let start = 1600000000000;
  while (true) {
    yield start;
    start += 100;
  }
}

let getTime = timeGenerator();

jest
  .spyOn(DateTime, "now")
  .mockImplementation(() => DateTime.fromMillis(getTime.next().value));

describe("DebugHistoryMiddleware", () => {
  let q = quark("FOO", {
    middlewares: [createDebugHistoryMiddleware({ name: "q1", trace: false })],
  });

  beforeEach(() => {
    getTime = timeGenerator();
    q.set("FOO");
    global?.__quark_history_tracker__?.clear();
  });

  it("correctly saves updates to the history tracker", () => {
    expect(q.get()).toEqual("FOO");

    q.set("BAR");

    expect(q.get()).toEqual("BAR");

    q.set(() => "BAZ");

    expect(q.get()).toEqual("BAZ");

    expect(global.__quark_history_tracker__).toBeDefined();

    expect(global.__quark_history_tracker__.showHistory()).toMatchSnapshot();
  });

  it("correctly handles async updates", async () => {
    expect(global.__quark_history_tracker__.showHistory()["0_q1"].length).toEqual(0);
    expect(q.get()).toEqual("FOO");

    q.set(Promise.resolve("BAR"));

    q.set(() => Promise.resolve("BAZ"));

    await sleep(0);

    expect(global.__quark_history_tracker__.showHistory()).toMatchSnapshot();
  });
});
