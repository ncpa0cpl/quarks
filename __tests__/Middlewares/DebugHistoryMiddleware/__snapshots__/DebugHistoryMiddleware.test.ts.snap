// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DebugHistoryMiddleware correctly handles async updates 1`] = `
{
  "0_q1": [
    {
      "change": "Postponed",
      "dispatchedUpdate": {
        "type": "Promise",
        "value": Promise {
          Symbol(__quark_internal_is_promise_canceled__): true,
        },
      },
      "initialState": {
        "type": "Value",
        "value": "FOO",
      },
      "source": "Sync-Dispatch",
      "stackTrace": undefined,
      "stateAfterUpdate": "FOO",
      "time": 1600000000100,
    },
    {
      "change": "Postponed",
      "dispatchedUpdate": {
        "type": "Generator",
        "value": [Function],
      },
      "initialState": {
        "type": "Value",
        "value": "FOO",
      },
      "source": "Sync-Dispatch",
      "stackTrace": undefined,
      "stateAfterUpdate": "FOO",
      "time": 1600000000200,
    },
    {
      "change": "Postponed",
      "dispatchedUpdate": {
        "type": "Promise",
        "value": Promise {
          Symbol(__quark_internal_is_promise_canceled__): false,
        },
      },
      "initialState": {
        "type": "Value",
        "value": "FOO",
      },
      "source": "Sync-Dispatch",
      "stackTrace": undefined,
      "stateAfterUpdate": "FOO",
      "time": 1600000000300,
    },
    {
      "change": "Immediate",
      "dispatchedUpdate": {
        "type": "Value",
        "value": "BAR",
      },
      "initialState": {
        "type": "Value",
        "value": "FOO",
      },
      "isCanceled": true,
      "source": "Async-Dispatch",
      "stackTrace": undefined,
      "stateAfterUpdate": "BAR",
      "time": 1600000000400,
    },
    {
      "change": "Immediate",
      "dispatchedUpdate": {
        "type": "Value",
        "value": "BAZ",
      },
      "initialState": {
        "type": "Value",
        "value": "FOO",
      },
      "source": "Async-Dispatch",
      "stackTrace": undefined,
      "stateAfterUpdate": "BAZ",
      "time": 1600000000500,
    },
  ],
}
`;

exports[`DebugHistoryMiddleware correctly saves updates to the history tracker 1`] = `
{
  "0_q1": [
    {
      "change": "Immediate",
      "dispatchedUpdate": {
        "type": "Value",
        "value": "BAR",
      },
      "initialState": {
        "type": "Value",
        "value": "FOO",
      },
      "source": "Sync-Dispatch",
      "stackTrace": undefined,
      "stateAfterUpdate": "BAR",
      "time": 1600000000100,
    },
    {
      "change": "Postponed",
      "dispatchedUpdate": {
        "type": "Generator",
        "value": [Function],
      },
      "initialState": {
        "type": "Value",
        "value": "BAR",
      },
      "source": "Sync-Dispatch",
      "stackTrace": undefined,
      "stateAfterUpdate": "BAR",
      "time": 1600000000200,
    },
    {
      "change": "Immediate",
      "dispatchedUpdate": {
        "type": "Value",
        "value": "BAZ",
      },
      "initialState": {
        "type": "Value",
        "value": "BAR",
      },
      "source": "Sync-Dispatch",
      "stackTrace": undefined,
      "stateAfterUpdate": "BAZ",
      "time": 1600000000300,
    },
  ],
}
`;
